packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -462,45 +229,21 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
     BuildContext context,
     Rect globalEditableRegion,
     double textLineHeight,
-    Offset position,
+    Offset selectionMidpoint,
     List<TextSelectionPoint> endpoints,
     TextSelectionDelegate delegate,
     ClipboardStatusNotifier clipboardStatus,
   ) {
-    assert(debugCheckHasMediaQuery(context));
-    final MediaQueryData mediaQuery = MediaQuery.of(context);
-
-    // The toolbar should appear below the TextField when there is not enough
-    // space above the TextField to show it, assuming there's always enough space
-    // at the bottom in this case.
-    final double toolbarHeightNeeded = mediaQuery.padding.top
-      + _kToolbarScreenPadding
-      + _kToolbarHeight
-      + _kToolbarContentDistance;
-    final double availableHeight = globalEditableRegion.top + endpoints.first.point.dy - textLineHeight;
-    final bool isArrowPointingDown = toolbarHeightNeeded <= availableHeight;
-
-    final double arrowTipX = (position.dx + globalEditableRegion.left).clamp(
-      _kArrowScreenPadding + mediaQuery.padding.left,
-      mediaQuery.size.width - mediaQuery.padding.right - _kArrowScreenPadding,
-    );
-
-    // The y-coordinate has to be calculated instead of directly quoting position.dy,
-    // since the caller (TextSelectionOverlay._buildToolbar) does not know whether
-    // the toolbar is going to be facing up or down.
-    final double localBarTopY = isArrowPointingDown
-      ? endpoints.first.point.dy - textLineHeight - _kToolbarContentDistance - _kToolbarHeight
-      : endpoints.last.point.dy + _kToolbarContentDistance;
-
-    return _CupertinoTextSelectionToolbarWrapper(
-      arrowTipX: arrowTipX,
-      barTopY: localBarTopY + globalEditableRegion.top,
+    return _CupertinoTextSelectionControlsToolbar(
```

Comment:
I moved all of the deleted math above into this class here to be consistent with Material.  I'd like to refactor this in the long term (see https://github.com/flutter/flutter/issues/73574), but for now the idea is to encapsulate the complexity and keep the public APIs nice and clean.